### ISSUE TITLE
Add possible caveats about usage of PyPI by linux distros

### ISSUE THREAD

**Issue by aragilar** [pypackaging-native/pypackaging-native#21]
It looks like sometimes linux distros do not use PyPI sdists as the source for their packages, and instead use other artefacts (e.g. git repos), due to the provided sdists not meeting (see the thread starting at https://lists.debian.org/debian-python/2020/10/msg00007.html). It might be worth adding that different users want different "source files" (i.e. linux distros strongly avoid vendoring/pre-generated files, whilst including the source of other libraries or pre-generating C code from Cython makes it easier for naive users to build from source) to the different purposes of PyPI page.

**Comment by h-vetinari** [comment]
Yeah, vendored pre-Cythonized files are an issue, as is the fact that sdists can diverge arbitrarily from the git tag they are derived from (-> possible attack vector).

For me builds instructions should always start from a vanilla git tag. If stuff needs to be generated, make it part of the build instructions. Of course, certain projects try to shield their users from the complexity that this can entail, but I don't think packages should distribute an inferior product to help naïve users. Thankfully, sdists are not the main distribution mechanism.

In any case, a short description of the current state of affairs of this would probably fit on the PyPI social model page.

**Comment by rgommers** [comment]
Great point, let's add that. I agree that adding generated Cython code is a bad practice by now (it used to be necessary 5-10 years ago). An sdist should ideally contain zero generated files, or a single one containing the git hash if that's used by dynamic version generation tools (e.g., like so https://github.com/FFY00/trampolim/commit/2c6f7c49c8dbbbe23e00ca6e9a09dfe734b60623 in order to do that automatically with `git-archive`).



> In any case, a short description of the current state of affairs of this would probably fit on the PyPI social model page.

Probably better on https://pypackaging-native.github.io/meta-topics/purposes_of_pypi/?

**Comment by h-vetinari** [comment]
> > In any case, a short description of the current state of affairs of this would probably fit on the PyPI social model page.
> 
> Probably better on https://pypackaging-native.github.io/meta-topics/purposes_of_pypi/?

Sure, sorry.

**Comment by rgommers** [comment]
> (see the thread starting at https://lists.debian.org/debian-python/2020/10/msg00007.html)

I now read the whole thread, and have to say that I expected something a bit more conclusive. It seems like there's no clear policy in Debian to prefer the original VCS tags over PyPI sdists (at least, this thread doesn't seem to say so)?

**Comment by eli-schwartz** [comment]
It depends a lot on the specific project. Also the distro.

PyPI sdists *can* be a perfectly valid release tarball. In your class C/C++ project with its library, CLI programs, GUI, desktop files, manpages, etc. the rule is exactly the same. Let's assume it's an autotools project -- common enough. There's a well defined routine for releasing a project, and it's called `make distcheck`. This figures out exactly what files need to be distributed inside a tarball release, makes a tarball, extracts the tarball, tries building the tarball and running its full suite of tests, and installs it too. If all that works, it says "yay it worked" and hands the user a tarball suitable for uploading places like GitHub releases.

Usefully, that tarball also contains generated configure scripts. This makes it vastly preferable to most distros, and also most *people* in general. Also usefully, that tarball does not contain stuff like the .github directory, which is used in CI and never by downstream packagers. It also doesn't include, well, any other files that aren't needed by downstream packagers. It ***does*** include the testsuite, and if there are docs, it invariably includes the docs too. Often pregenerated copies of the docs.

This works quite well because there are best practices for this sort of thing, and autotools automatically does the work for you -- if the file is used in the build system, the build system packages it up at dist time. (Unfortunately, if the build system doesn't *obviously* use it you need to handhold it quite a bit, this is for example why you must list headers as sources for a target even though headers aren't actually built. It's literally documented in autotools that this doesn't really do anything other than get dist to work.)

When it comes to python ecosystem packaging, this is unfortunately a bit of a disaster. There's no standard sdist process. Lots of people in the python community are used to sdist being the only thing uploaded to PyPI, and they don't like extra kilobytes of download when installing something, so they don't do the obvious thing and push for compression methods a bit more modern than the venerable and ubiquitous DEFLATE. Instead, they strip out all the things that downstream packagers consider vital, like testsuites, documentation, license files (????????) etc. In self defense, downstream packagers then resort to what actually works for their purposes, i.e. GitHub autogenerated tarballs.

Compounding the problem, some members of PyPA state their general unbelief in the idea of shipping tests in sdists, e.g. because "pip doesn't do anything with them anyway", and then other projects see that, say "well, if this is what PyPA's official stance is, it must be best practice to not include tests. I should follow best practices and not do so either". It's of course not an official PyPA stance, but how obvious is that nuance to someone that merely sees the speaker is a PyPA member?

There are lots of well behaved python projects that have good quality sdists too. So some distros have a policy to use sdists where possible, and GitHub where not possible. Other distros shun sdists on general principle, throwing the baby out with the bathwater.

**Comment by eli-schwartz** [comment]
On the topic of pre-generating cython files, ideally the build system offers some mechanism for regenerating that. And some distros do also want to regenerate autotools configure scripts, for... pretty good reason, actually, because they support uncommon architectures and need the latest fixes to autotools to handle that well. Fortunately this is trivial to do when the dist tarball already has it -- just run `autoreconf -fi`. Other build systems that don't embed the build system into the dist tarball don't have this issue.

It's not generally a reason to avoid high quality dist tarballs (python sdist or otherwise).

**Comment by rgommers** [comment]
Thanks for all the context @eli-schwartz!

> Instead, they strip out all the things that downstream packagers consider vital, like testsuites, documentation, license files 

I believe a big part of the problem here is that there's no good way with distutils/setuptools to say "include in sdist, but don't install". What Flit, meson-python & co do (use everything tracked by git to include in the sdist) avoids that naturally.

**Comment by eli-schwartz** [comment]
Isn't that just "list in MANIFEST.in, and that is all"?

setuptools_scm also does this for free.

**Comment by rgommers** [comment]
> Isn't that just "list in MANIFEST.in, and that is all"?

Not quite. I forgot all the ugly details, but `distutils` and `setuptools` handle MANIFEST.in differently, and a very quick check shows that numpy installs have `setup.py` files floating around in the install dir rather than just in the sdist. Maybe that's all a legacy problem, not sure what the status is with latest setuptools version only.

**Comment by eli-schwartz** [comment]
Intriguing. I thought that the only meaningful difference to a wheel, if MANIFEST.in is used, is that `include_package_data=True` assumes that all files in the sdist are package data that you want to add to the wheel. But it's been a while since I touched plain distutils (something that's not broadly useful except in python <=3.11 in combination with solving the setuptools bootstrap topic).

(Is this about the root setup.py or the numpy extension of using subdir setup.py files?)

**Comment by rgommers** [comment]
> (Is this about the root setup.py or the numpy extension of using subdir setup.py files?)

root and subdir ones are all present.
